diff --git a/java/java.editor/src/org/netbeans/modules/editor/java/JavaStructureProvider.java b/java/java.editor/src/org/netbeans/modules/editor/java/JavaStructureProvider.java
index 61d08c667131..be87f4b2b830 100644
--- a/java/java.editor/src/org/netbeans/modules/editor/java/JavaStructureProvider.java
+++ b/java/java.editor/src/org/netbeans/modules/editor/java/JavaStructureProvider.java
@@ -65,10 +65,13 @@ public List<StructureElement> getStructure(Document doc) {
                         TreePath tp = trees.getPath(cu, cu.getPackage());
                         Element el = trees.getElement(tp);
                         if (el != null && el.getKind() == ElementKind.PACKAGE) {
-                            StructureElement jse = element2StructureElement(cc, el);
-                            if (jse != null) {
-                                result.add(jse);
-                            }
+                            Builder builder = StructureProvider.newBuilder(el.getSimpleName().toString(), ElementHeaders.javaKind2Structure(el));
+                            int start = (int) cc.getTrees().getSourcePositions().getStartPosition(cu, cu.getPackage());
+                            int end = (int) cc.getTrees().getSourcePositions().getEndPosition(cu, cu.getPackage());
+
+                            builder.expandedStartOffset(start).selectionStartOffset(start);
+                            builder.expandedEndOffset(end).selectionEndOffset(end);
+                            result.add(builder.build());
                         }
                     }
                     for (Element tel : cc.getTopLevelElements()) {
