diff --git a/nbbuild/misc/prepare-bundles/src/main/java/org/netbeans/prepare/bundles/PrepareBundles.java b/nbbuild/misc/prepare-bundles/src/main/java/org/netbeans/prepare/bundles/PrepareBundles.java
index 5349a20e09..536df96b0e 100644
--- a/nbbuild/misc/prepare-bundles/src/main/java/org/netbeans/prepare/bundles/PrepareBundles.java
+++ b/nbbuild/misc/prepare-bundles/src/main/java/org/netbeans/prepare/bundles/PrepareBundles.java
@@ -69,7 +69,12 @@ public class PrepareBundles {
 
         Path targetDir = Paths.get(args[0]);
         Path packagesDir = targetDir.resolve("package");
-        new ProcessBuilder("npm", "install").directory(packagesDir.toFile()).inheritIO().start().waitFor();
+        String os = System.getProperty("os.name").toLowerCase();
+        if (os.contains("windows")) {
+            new ProcessBuilder("npm.cmd", "install").directory(packagesDir.toFile()).inheritIO().start().waitFor();
+        } else{
+            new ProcessBuilder("npm", "install").directory(packagesDir.toFile()).inheritIO().start().waitFor();   
+        }
         Path bundlesDir = targetDir.resolve("bundles");
         Files.createDirectories(bundlesDir);
         try (DirectoryStream<Path> ds = Files.newDirectoryStream(bundlesDir)) {
@@ -107,18 +112,16 @@ public class PrepareBundles {
                 if ("@types".equals(module.getFileName().toString())) continue;
                 if ("@esbuild".equals(module.getFileName().toString())) continue;
                 if ("@microsoft".equals(module.getFileName().toString())) continue;
-                if ("@vscode".equals(module.getFileName().toString())) {
-                    try (DirectoryStream<Path> sds = Files.newDirectoryStream(module)) {
-                        for (Path sModule : sds) {
-                            checkModule(sModule, sb, tokens2Projects, project2License, bundlesDir, targetDir, externalDir, binariesList);
-                        }
-                    }
-                    continue;
+                Path packageJson = module.resolve("package.json");
+                if (Files.isReadable(packageJson)) {
+                    checkModule(module, sb, tokens2Projects, project2License, bundlesDir, targetDir, externalDir, binariesList);
+                    continue;    
                 }
-                if ("@ungap".equals(module.getFileName().toString())) {
-                    module = module.resolve("promise-all-settled");
+                try (DirectoryStream<Path> sds = Files.newDirectoryStream(module)) {
+                    for (Path sModule : sds) {
+                        checkModule(sModule, sb, tokens2Projects, project2License, bundlesDir, targetDir, externalDir, binariesList);
+                    }
                 }
-                checkModule(module, sb, tokens2Projects, project2License, bundlesDir, targetDir, externalDir, binariesList);
             }
         }
         if (sb.length() > 0) {
