diff --git a/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractJavaPlatformProviderOverride.java b/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractJavaPlatformProviderOverride.java
index df668954dd..054f85d0e5 100644
--- a/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractJavaPlatformProviderOverride.java
+++ b/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui/AbstractJavaPlatformProviderOverride.java
@@ -39,6 +39,14 @@ public abstract class AbstractJavaPlatformProviderOverride implements JavaPlatfo
     private final PropertyChangeSupport pcs = new PropertyChangeSupport(this);
     private FileObject defaultPlatformOverride;
 
+    public AbstractJavaPlatformProviderOverride() {
+        this(null);
+    }
+
+    protected AbstractJavaPlatformProviderOverride(String defaultPlatformOverride) {
+        this.defaultPlatformOverride = defaultPlatformOverride == null ? null : FileUtil.toFileObject(new File(defaultPlatformOverride));
+    }
+
     @Override
     @SuppressWarnings("ReturnOfCollectionOrArrayField")
     public JavaPlatform[] getInstalledPlatforms() {
