diff --git a/ide/editor.lib2/src/org/netbeans/spi/editor/highlighting/support/HighlightsContainers.java b/ide/editor.lib2/src/org/netbeans/spi/editor/highlighting/support/HighlightsContainers.java
index 5447ab950347..56de41045d8d 100644
--- a/ide/editor.lib2/src/org/netbeans/spi/editor/highlighting/support/HighlightsContainers.java
+++ b/ide/editor.lib2/src/org/netbeans/spi/editor/highlighting/support/HighlightsContainers.java
@@ -38,6 +38,9 @@ class InlineHintsSettingsAwareContainer extends AbstractHighlightsContainer impl
 
             public InlineHintsSettingsAwareContainer() {
                 String mimeType = DocumentUtilities.getMimeType(doc);
+                if (mimeType == null) {
+                    mimeType = "";
+                }
                 prefs = MimeLookup.getLookup(mimeType).lookup(Preferences.class);
                 prefs.addPreferenceChangeListener(WeakListeners.create(PreferenceChangeListener.class, this, prefs));
                 inlineSettingChanged();
