diff --git a/java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java b/java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java
index cacf2c8d43e1..c9c4a4417350 100644
--- a/java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java
+++ b/java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java
@@ -365,6 +365,7 @@ private String searchLocation(Testcase tc, String className, String methodName,
 
         String relativePath = null;
         for (Map.Entry<ClasspathInfo, Path> ci : classpathInfo.entrySet()) {
+            if (ci.getKey() == null) continue;
             FileObject fo = SourceUtils.getFile(ElementHandle.createTypeElementHandle(ElementKind.CLASS, className), ci.getKey());
             if (fo != null) {
                 relativePath = ci.getValue().relativize(FileUtil.toFile(fo).toPath()).toString();
