diff --git a/extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java b/extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java
index f797c28ce570..7a62d2d8dd2e 100644
--- a/extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java
+++ b/extide/gradle/netbeans-gradle-tooling/src/main/java/org/netbeans/modules/gradle/tooling/NbProjectInfoBuilder.java
@@ -1705,7 +1705,7 @@ private void detectDependencies(NbProjectInfoModel model) {
             model.registerPerf(depPrefix + "project", time_file_deps - time_project_deps);
             Set<File> fileDeps = new HashSet<>();
             it.getDependencies().withType(FileCollectionDependency.class).forEach(it2 -> {
-                fileDeps.addAll(it2.resolve());
+                it2.getFiles().forEach(fileDeps::add);
             });
             long time_collect = System.currentTimeMillis();
             model.registerPerf(depPrefix + "file", time_collect - time_file_deps);
