diff --git a/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java b/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java
index eaff1013cd..97db75f5af 100644
--- a/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java
+++ b/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java
@@ -1422,6 +1422,17 @@ public final class WorkspaceServiceImpl implements WorkspaceService, LanguageCli
             } catch (IOException ex) {
                 Exceptions.printStackTrace(ex);
             }
+            // When no path set or wanting to override specific option
+            Preferences prefs = CodeStylePreferences.get(fo, "text/x-java").getPreferences();
+            JsonObject formatOptions = configuration.getAsJsonObject("options");
+            if (formatOptions != null) {
+                Map<String, JsonElement> options = formatOptions.asMap();
+                for (String key : options.keySet()) {
+                    if (options.get(key) != null) {
+                        prefs.put(key, options.get(key).getAsString());
+                    }
+                }
+            }
         }
     }
 
